src/query/api/v1/middleware/rewrite.go

```diff
@@ -183,12 +196,73 @@ func rewriteRangeDuration(
 	logger.Debug("rewrote duration values in request",
 		zap.String("originalQuery", params.query),
 		zap.String("updatedQuery", updatedQuery),
-		zap.Duration("originalLookback", params.lookback),
+		zap.Duration("originalLookback", originalLookback),
 		zap.Duration("updatedLookback", updatedLookback))
 
 	return nil
 }
 
+func findLargestQueryResolution(ctx context.Context,
+	store storage.Storage,
+	fetchOpts *storage.FetchOptions,
+	startTime time.Time,
+	endTime time.Time,
+) (time.Duration, error) {
+	attrs, err := store.QueryStorageMetadataAttributes(ctx, startTime, endTime, fetchOpts)
+	if err != nil {
+		return 0, err
+	}
+
+	// Find the largest resolution
+	var res time.Duration
+	for _, attr := range attrs {
+		if attr.Resolution > res {
+			res = attr.Resolution
+		}
+	}
+	return res, nil
+}
+
+// Using the prometheus engine in this way should be considered
+// optional and best effort. Fall back to the frequently accurate logic
+// of using the start and end time in the request
+func getQueryBounds(
+	opts PrometheusRangeRewriteOptions,
+	params params,
+	fetchOpts *storage.FetchOptions,
+	logger *zap.Logger,
+) (start time.Time, end time.Time) {
+	start = params.start
+	end = params.end
+
+	if opts.PrometheusEngineFn != nil {
```

Comment:
nit: Another Go idiom is preferring early return. So this should look like

```go
if opts.PrometheusEngineFn == nil {
  return param.start, params.end
}

// Do rest
```
In addition to being more idiomatic, it reduces nesting which makes cod easier to read